Add a way to use a custom webpack config

vcua-mobify opened the pull request:
Small change to allow projects to use a custom webpack.config.js that is placed in the root of the pwa.

vcua-mobify committed:
Add a way to use a custom webpack config

vcua-mobify committed:
Add a way to use a custom webpack config

vcua-mobify commented:
Another small change could be to add a small template webpack.config.js file in /pwa that just imports and exports the default configs so users don't have to create the file.

stm-dschulz committed:
Fix/258 forgotpassword not implemented in checkout 2 (#373)

vmarta committed:
Footer: locale selector's options no longer appear blank on Windows (#381)

adamraya committed:
[Generator] Change generator questions (#374)

* Update generation question copy

* Slugify Project Name

* Update packages/pwa-kit-create-app/scripts/create-mobify-app.js

Co-authored-by: John Boxall <jboxall@salesforce.com>

* lint

* PR Feedback

* Remove Einstein question

* PR Feedback regex dynamic max length, Re-use validProjectName

* Use projectId as outputDir

* Clean up

Co-authored-by: John Boxall <jboxall@salesforce.com>

vmarta committed:
Show warning when using incompatible node version (#384)

Main changes are:
- Making sure that all packages have defined and use consistent engines field in their package.json
- Within our generator script, perform a check for the developers' node version and show a warning if it's incompatible

vcua-mobify committed:
Import cross-fetch to make OCAPI client isomorphic (#382)

* Import cross-fetch to make OCAPI client isomorphic

* mock cross fetch

* Fix remaining tests that use ocapi client

* Downgrade version of jest-fetch-mock to match the rest of the monorepo

* Add the carat to the version

* Only mock cross-fetch in the required tests

* Set global.fetch in jest setup

* Fix polyfill test

* Remove dead class with a broken test

* Move jest-fetch-mock setup to individual tests.

* Fix tests

Co-authored-by: Kevin He <kevin.he@salesforce.com>

vcua-mobify committed:
Fix lint error

vcua-mobify committed:
Fix lint error

vcua-mobify commented:
@olibrook when you get the chance, could you review this change to add a way to use a custom webpack config to the projects generated from bfh?

olibrook commented:
@vcua-mobify This looks like a sensible change. Can I ask you to show me it working, somehow? One thing I'm concerned about is how webpack looks up plugins, etc. Can you show me an example of extending our config and adding some new plugin to it in the project? With it all working?!

vcua-mobify commented:
@olibrook I've opened a separate PR #401 just for testing the use of a custom webpack config.

Currently, I am finding that `npm start` is not picking up the custom config I've created so there's probably more changes required in this PR to get that working.

vmarta committed:
Fix peerDependencies that aren't updated with bump-version

Similar fix like this previous commit: https://github.com/SalesforceCommerceCloud/pwa-kit/commit/1329dd3bb8bfebc3d25db3734afa2577f5e4ec75

alexvuong committed:
remove url-parse as it is not longer needed (#411)

Commit:
Describe mrt usage inside NPM scripts, to avoid common errors with arg-forwaring.

Commit:
Merge branch 'black-friday-hackathon' of https://github.com/SalesforceCommerceCloud/pwa-kit into black-friday-hackathon

Commit:
Show version, format

vcua-mobify committed:
Use a top level webpack config to import custom configs

vcua-mobify committed:
Use a top level webpack config to import custom configs

vcua-mobify commented:
I've changed the approach used to load a custom webpack. Rather than modifying mrt.js, I've added a top level webpack config that will import a webpack.config.js in the project directory if it exists. If it does not exist, it will use the default config in pwa-kit-build.

PR #401 has been updated to show this working by adding css-loader and style-loader via custom config

vmarta committed:
Tweak formatting of the options documentation

For consistency with other options, separate the short and long versions with comma.

Commit:
Merge branch 'develop' into black-friday-hackathon

Commit:
Format

Commit:
Delete all redundant scripts in react-sdk - Vincent's feedback

Commit:
Delete unused archiver code and unused react-sdk deps.

Commit:
Remove unused deps and upload code

Commit:
Delete Utils.errorForStatus which is unused

Commit:
Delete Utils.readCredentials which is unused

Commit:
Delete Utils.requestErrorMessage which is unused

Commit:
Delete all unused code in scripts/utils.js and dependencies

Commit:
Vet/prune all SDK dependencies - Vincent's feedback

Commit:
Format

Commit:
Fix a whole bunch of tests

Commit:
Fix tests

vmarta committed:
Fix bug with extending an options object (#419)

vmarta committed:
Propage the engines field to all the other packages

vmarta committed:
Merge branch 'develop' into black-friday-hackathon

vmarta committed:
Remove stale .nvmrc

It looks like we haven't really kept these files up to date in V1.

We now supports both Node 12 and 14, while .nvmrc file is supposed to point to a single node version. I think it's best to let others decide which version they'd like to use.

kevinxh committed:
remove config validation webpack plugin (#443)

kevinxh committed:
[V2] Remove `body-parser` middleware (#444)

* remove bodyparser

* remove legacy tests

vcua-mobify committed:
[V2] Remove deprecations (#425)

* Remove create-hash-manifest script

* Remove legacy remote proxying

* Ignore unused var on proxy function

* Ignore unused var on another proxy function

* Remove create-hash-manifest script

* Remove legacy remote proxying

* Ignore unused var on proxy function

* Ignore unused var on another proxy function

* Update changelogs

adamraya committed:
Merge `develop` into bfh 20220302 (#434)

* Fix SDK's peer dependencies versioning (#432)

Previously the range was matching all possible versions. Now matching React 16 and 17 only.

* Node warnings with better version formatting (#410)

The node version range is now taken from the engines field of the package.json, which is more readable than seeing semver's representation (>=12.0.0 <13.0.0-0||>=14.0.0 <15.0.0-0).

* [Circle CI] Add test generated project job on Windows (#412)

* Update config.yml

* Add `generatedSFCCProjectTestWindows` test job

* Test not using gtime.js script

* Increase log verbosity

* Run first the CI job generated project on Windows

* Use  outputDir generated-project

* Use execSync

* Update npx pwa-kit-create-app command

* Use process.argv as outputDir

* Test execFileSync

* Add try/catch to execSync

* Show npm i logging

* Show npm install logs as we do in the BFH branch

* Show verdaccio logs

* More logs npm install verbose

* Use Node child_process & Update verdaccio config & bump version

* Use verdaccio npmjs cache true

* Set GENERATOR_PRESET env variable

* Update create-mobify-app.js

* Test using a xlarge Windows machine

* Increase Verdaccio max_fails config

* Remove server keepAliveTimeout

* Dump all Verdaccio logs to a file for debugging

* Fix smoketestscripts on Windows

* Add Circle CI unit to the scripts path

* Really fix smoketestscripts Circle CI job on Windows

* Update config.yml

* More testing smoke test script on Windows

* And more testing smoke test script on Windows

* smoketestscripts job working on Windows

* fix typo

* Clean up

* More clean up

* Update latest v4 verdaccio

* Remove file from bad merge

Co-authored-by: Vincent Marta <vmarta@salesforce.com>

adamraya commented on packages/pwa-kit-build/src/configs/webpack/config.js:
```diff
@@ -8,325 +8,14 @@
-const findInProjectThenSDK = (pkg) => {
-    const projectPath = resolve(projectDir, 'node_modules', pkg)
-    return fs.existsSync(projectPath) ? projectPath : resolve(sdkDir, 'node_modules', pkg)
+if (fs.existsSync(projectWebpackPath)) {
+    module.exports = require(projectWebpackPath)
+} else {
+    module.exports = require('./base-config.js')
```
Suppose users completely replace the Webpack configuration we provide. In that case, they will need to copy and paste all the Webpack configuration that we require into the project to get things working again and then customize it. 

Once they choose to use their own Webpack configuration file in the project, users will completely control the Webpack configuration. But, at the same time, users will no longer benefit from the updates that we provide to our Webpack configuration.

In my opinion, this is not the best API experience to customize the Webpack configuration. I think the experience provided in V1 has some advantages that we should consider.

V1 allows users to override parts of the configuration or extend the Webpack configuration. This approach ensures that users consume future updates and that the configuration is less likely to break. It has the disadvantage that they need to use our Webpack configuration.

@vcua-mobify Are you open to summarizing the possible approaches in a Technical proposal doc and discussing it with the team during our Technical discussion meeting before deciding how we want to allow users to customize the Webpack configuration in V2?

bendvc replied:
I'd have to agree with @adamraya on this. You approach almost seems like a partial implementation of `create-react-app`'s eject feature, minus we aren't providing the automation of ejecting. I can't say I love either our approach or the new one, but the existing V1 behaviour is the lesser of 2 evils in my eyes.

bendvc committed:
[V2] Port Environment Specific Configuration Support (#447)

* Migrate Environment Specific Configuration Support

vmarta commented on packages/pwa-kit-build/src/configs/webpack/config.js:
```diff
@@ -8,325 +8,14 @@
-    }
-    return false
-}
+const projectWebpackPath = path.resolve(projectDir, 'webpack.config.js')
```
@vcua-mobify Currently our pwa template does _not_ have this webpack config file. But let's reconsider that, because the template does have other config files, even when some are only one-liner. I think to be consistent (and also easier to discover by the end users), we need to create this webpack.config.js in the pwa too.

The pwa template currently has these other configs, which has a pattern of importing from `pwa-kit-build`:
- [babel.config.js](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/6be1fc698ce9a9df58277cd58e45f66accd4b098/packages/pwa/babel.config.js#L7) - simply import
- [jest.config.js](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/6be1fc698ce9a9df58277cd58e45f66accd4b098/packages/pwa/jest.config.js#L1-L8) - import and also extending the base config

vcua-mobify committed:
Revert "Use a top level webpack config to import custom configs"

This reverts commit 980ae2efae7a2654cd098f4dd1a8be081285a04c.

vcua-mobify committed:
Conditionally import webpack from project in the dev server

vcua-mobify committed:
Add webpack.config.js customization points in apps

vcua-mobify committed:
Merge branch 'bfh-custom-webpack' of github.com:SalesforceCommerceCloud/pwa-kit into bfh-custom-webpack

vcua-mobify committed:
Delete base-config.js

vcua-mobify commented:
Git commit history has gotten messy after I tried to rebase with the black-friday-hackathon branch. I'll reopen an new PR with the git history cleaned up.

vcua-mobify commented:
New PR #462 has been created.
